File: java/Dson-Core/src/main/java/cn/wjybxx/dson/internal/CommonsLang3.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package cn.wjybxx.dson.internal;

import cn.wjybxx.base.ObjectUtils;
import cn.wjybxx.dson.io.DsonIOException;

/**
 * 这里的算法修改自commons-lang3，commons-lang3现在文件太多，不想引入。
 *
 * @author wjybxx
 * date - 2023/12/16
 */
public class CommonsLang3 {

    // region number
    public static boolean isParsable(String str) {
        if (ObjectUtils.isEmpty(str)) {
            return false;
        }
        if (str.charAt(str.length() - 1) == '.') {
            return false;
        }
        if (str.charAt(0) == '-') {
            if (str.length() == 1) {
                return false;
            }
            return withDecimalsParsing(str, 1);
        }
        return withDecimalsParsing(str, 0);
    }

    private static boolean withDecimalsParsing(final String str, final int beginIdx) {
        int decimalPoints = 0;
        for (int i = beginIdx; i < str.length(); i++) {
            final boolean isDecimalPoint = str.charAt(i) == '.';
            if (isDecimalPoint) {
                decimalPoints++;
            }
            if (decimalPoints > 1) {
                return false;
            }
            if (!isDecimalPoint && !Character.isDigit(str.charAt(i))) {
                return false;
            }
        }
        return true;
    }
    // endregion

    // region hex
    private static final char[] DIGITS_UPPER = {
            '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
            'A', 'B', 'C', 'D', 'E', 'F'};

    public static void encodeHex(final byte[] data, final int dataOffset, final int dataLen,
                                 final char[] outBuffer, final int outOffset) {
        char[] toDigits = DIGITS_UPPER;
        for (int i = dataOffset, j = outOffset; i < dataOffset + dataLen; i++) {
            outBuffer[j++] = toDigits[(0xF0 & data[i]) >>> 4]; // 高4位
            outBuffer[j++] = toDigits[0x0F & data[i]]; // 低4位
        }
    }

    public static byte[] decodeHex(char[] data) {
        final int dateLen = data.length;
        if ((dateLen & 0x01) != 0) {
            throw new DsonIOException("string length not even: " + data.length);
        }
        byte[] result = new byte[dateLen >> 1];
        // two characters form the hex value.
        for (int i = 0, j = 0; j < dateLen; i++) {
            int f = toDigit(data[j], j) << 4;
            j++;
            f = f | toDigit(data[j], j);
            j++;
            result[i] = (byte) (f & 0xFF);
        }
        return result;
    }

    private static int toDigit(final char c, final int index) {
        return switch (c) {
            case '0' -> 0;
            case '1' -> 1;
            case '2' -> 2;
            case '3' -> 3;
            case '4' -> 4;
            case '5' -> 5;
            case '6' -> 6;
            case '7' -> 7;
            case '8' -> 8;
            case '9' -> 9;
            case 'a', 'A' -> 10;
            case 'b', 'B' -> 11;
            case 'c', 'C' -> 12;
            case 'd', 'D' -> 13;
            case 'e', 'E' -> 14;
            case 'f', 'F' -> 15;
            default -> throw new DsonIOException("Illegal hexadecimal character " + c + " at index " + index);
        };
    }
    // endregion
}

File: csharp/Wjybxx.Dson.Core/src/Internal/CommonsLang3.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Text;
using Wjybxx.Commons.Pool;
using Wjybxx.Dson.IO;

namespace Wjybxx.Dson.Internal
{
/// <summary>
/// 这里的算法修改自Java的commons-lang3包
/// </summary>
internal static class CommonsLang3
{
    public static bool IsParsable(string str) {
        if (string.IsNullOrEmpty(str)) {
            return false;
        }
        if (str[str.Length - 1] == '.') {
            return false;
        }
        if (str[0] == '-') {
            if (str.Length == 1) {
                return false;
            }
            return WithDecimalsParsing(str, 1);
        }
        return WithDecimalsParsing(str, 0);
    }

    private static bool WithDecimalsParsing(string str, int beginIdx) {
        int decimalPoints = 0;
        for (int i = beginIdx; i < str.Length; i++) {
            bool isDecimalPoint = str[i] == '.';
            if (isDecimalPoint) {
                decimalPoints++;
            }
            if (decimalPoints > 1) {
                return false;
            }
            if (!isDecimalPoint && !char.IsDigit(str[i])) {
                return false;
            }
        }
        return true;
    }

    /** 字节数组转16进制 */
    private static readonly char[] DIGITS_UPPER = new[]
    {
        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
        'A', 'B', 'C', 'D', 'E', 'F'
    };

    public static string ToHexString(byte[] data) {
#if NET5_0_OR_GREATER
        return Convert.ToHexString(data);
#else
        char[] buffer = IArrayPool<char>.Shared.Acquire(data.Length * 2);
        try {
            EncodeHex(data, 0, data.Length, buffer);
            return new string(buffer, 0, data.Length * 2);
        }
        finally {
            IArrayPool<char>.Shared.Release(buffer);
        }
#endif
    }

    public static byte[] FromHexString(string hexString) {
#if NET5_0_OR_GREATER
        return Convert.FromHexString(hexString);
#else
        return DecodeHex(hexString);
#endif
    }

    /** 编码长度固定位 dataLen * 2 */
    public static void EncodeHex(byte[] data, int dataOffset, int dataLen,
                                 Span<char> outBuffer) {
        char[] toDigits = DIGITS_UPPER;
        for (int i = dataOffset, j = 0; i < dataOffset + dataLen; i++) {
            outBuffer[j++] = toDigits[(0xF0 & data[i]) >> 4]; // 高4位
            outBuffer[j++] = toDigits[0x0F & data[i]]; // 低4位
        }
    }

    public static byte[] DecodeHex(string hexString) {
        int dateLen = hexString.Length;
        if ((dateLen & 0x01) != 0) {
            throw new ArgumentException("string length not even: " + hexString.Length);
        }
        byte[] result = new byte[dateLen >> 1];
        // two characters form the hex value.
        for (int i = 0, j = 0; j < dateLen; i++) {
            int f = ToDigit(hexString[j], j) << 4;
            j++;
            f = f | ToDigit(hexString[j], j);
            j++;
            result[i] = (byte)(f & 0xFF);
        }
        return result;
    }

    public static byte[] DecodeHex(StringBuilder hexString) {
        int dateLen = hexString.Length;
        if ((dateLen & 0x01) != 0) {
            throw new ArgumentException("string length not even: " + hexString.Length);
        }
        byte[] result = new byte[dateLen >> 1];
        // two characters form the hex value.
        for (int i = 0, j = 0; j < dateLen; i++) {
            int f = ToDigit(hexString[j], j) << 4;
            j++;
            f = f | ToDigit(hexString[j], j);
            j++;
            result[i] = (byte)(f & 0xFF);
        }
        return result;
    }

    private static int ToDigit(char c, int index) {
        return c switch
        {
            '0' => 0,
            '1' => 1,
            '2' => 2,
            '3' => 3,
            '4' => 4,
            '5' => 5,
            '6' => 6,
            '7' => 7,
            '8' => 8,
            '9' => 9,
            //
            'a' => 10,
            'A' => 10,
            //
            'b' => 11,
            'B' => 11,
            //
            'c' => 12,
            'C' => 12,
            //
            'd' => 13,
            'D' => 13,
            //
            'e' => 14,
            'E' => 14,
            //
            'f' => 15,
            'F' => 15,
            _ => throw new DsonIOException("Illegal hexadecimal character " + c + " at index " + index)
        };
    }
}
}